### PR TITLE
ENG-6718 add registered user id support

### PR DESCRIPTION
**Pull request by kishisaka-nid**
Add support for registered user id. New methods are: 

```
fun setRegisteredUserID(registeredUserId: String): Boolean {
fun getRegisteredUserID(): String

```
JSON request body snippet: 
```
"type":"SET_REGISTERED_USER_ID","ts":1701279641179,"uid":"ahsdkghasdjkghdklasglasd"

"siteId":"form_picks709","userId":"1701279640946","clientId":"2cdc3c1b-7574-435c-a6e7-92f36a116cd5","identityId":"1701279640946","registeredUserId":"ahsdkghasdjkghdklasglasd"
```

### CODE REVIEW

File: NeuroID/src/main/java/com/neuroid/tracker/NeuroID.kt [hunk 1]
```diff
@@ -151,43 +151,69 @@ class NeuroID private constructor(
         return true
     }
 
-    fun setUserID(userId: String): Boolean {
-
-        val validUserID = this.validateUserId(userId)
-        if (!validUserID) {
-            return false
+    fun setRegisteredUserId(registeredUserId: String): Boolean {
```
Comment (kevinsites) [inline NeuroID/src/main/java/com/neuroid/tracker/NeuroID.kt:154]: can we do `setRegisteredUserID`? trying to keep the function names consistent across all 3 sdks

File: NeuroID/src/main/java/com/neuroid/tracker/NeuroID.kt [hunk 2]
```diff
@@ -151,43 +151,69 @@ class NeuroID private constructor(
...
+            return false
+        }
     }
 
 
     @Deprecated("Replaced with getUserID", ReplaceWith("getUserID()"))
     fun getUserId() = userID
     fun getUserID() = userID
 
+    fun getRegisteredUserId() = registeredUserID
```
Comment (kevinsites) [inline NeuroID/src/main/java/com/neuroid/tracker/NeuroID.kt:215]: same as above `getRegisteredUserID` please

File: NeuroID/src/main/java/com/neuroid/tracker/storage/NIDSharedPrefsDefaults.kt [hunk 3]
```diff
@@ -47,6 +47,16 @@ class NIDSharedPrefsDefaults(
         putString(NID_UID, userId)
     }
 
+    fun setRegisteredUserID(userId: String) {
+        putString(NID_REG_UID, userId)
+    }
+
+    fun getRegisteredUserId(): Any {
```
Comment (kevinsites) [inline NeuroID/src/main/java/com/neuroid/tracker/storage/NIDSharedPrefsDefaults.kt:54]: same as above

**Review by kevinsites — CHANGES_REQUESTED** [review]

File: NeuroID/src/main/java/com/neuroid/tracker/service/NIDServiceTracker.kt [hunk 4]
```diff
@@ -81,6 +81,7 @@ object NIDServiceTracker {
             put("userId", sharedDefaults.getUserId())
             put("clientId", sharedDefaults.getClientId())
             put("identityId", sharedDefaults.getUserId())
+            put("registeredUserID", sharedDefaults.getRegisteredUserId())
```
Comment (jeff-niemann) [inline NeuroID/src/main/java/com/neuroid/tracker/service/NIDServiceTracker.kt:84]: ID needs to be Id here
https://github.com/Neuro-ID/trackjs/blob/main/src/collection/event-queue.js#L170

File: NeuroID/src/main/java/com/neuroid/tracker/storage/NIDSharedPrefsDefaults.kt (hunk 3 above)
Comment (kishisaka-nid) [reply NeuroID/src/main/java/com/neuroid/tracker/storage/NIDSharedPrefsDefaults.kt:54]: These can stay the same and follow the existing convention in the file.

File: NeuroID/src/main/java/com/neuroid/tracker/NeuroID.kt (hunk 1 above)
Comment (kishisaka-nid) [reply NeuroID/src/main/java/com/neuroid/tracker/NeuroID.kt]: fixed. thanks!

File: NeuroID/src/main/java/com/neuroid/tracker/NeuroID.kt (hunk 2 above)
Comment (kishisaka-nid) [reply NeuroID/src/main/java/com/neuroid/tracker/NeuroID.kt]: fixed. thanks!

File: NeuroID/src/main/java/com/neuroid/tracker/storage/NIDSharedPrefsDefaults.kt (hunk 3 above)
Comment (kishisaka-nid) [reply NeuroID/src/main/java/com/neuroid/tracker/storage/NIDSharedPrefsDefaults.kt:54]: fixed. thanks!

File: NeuroID/src/main/java/com/neuroid/tracker/service/NIDServiceTracker.kt (hunk 4 above)
Comment (kishisaka-nid) [reply NeuroID/src/main/java/com/neuroid/tracker/service/NIDServiceTracker.kt]: fixed. thanks!

File: NeuroID/src/main/java/com/neuroid/tracker/storage/NIDSharedPrefsDefaults.kt [hunk 5]
```diff
@@ -47,7 +47,7 @@ class NIDSharedPrefsDefaults(
         putString(NID_UID, userId)
     }
 
-    fun setRegisteredUserID(userId: String) {
+    fun setRegisteredUserId(userId: String) {
```
[the remark this answers is not in the document]
Comment (kishisaka-nid) [reply NeuroID/src/main/java/com/neuroid/tracker/storage/NIDSharedPrefsDefaults.kt:50]: All of the above and below set Id methods use the "Id" convention so the setRegisteredUserId() should conform as well in this case. The NeuroID methods were fixed to "ID".

**Review by kevinsites — APPROVED** [review]

**Review by priya-xavier — DISMISSED** [review]

File: NeuroID/src/test/java/com/neuroid/tracker/NeuroIDClassUnitTests.kt [hunk 6]
```diff
@@ -265,6 +288,16 @@ class NeuroIDClassUnitTests {
         assertErrorCount(1)
     }
 
+    @Test
+    fun testSetRegisteredUserID_failure() {
+        setNeuroIDMockedLogger(errorMessage = "Invalid UserID")
+
+        val value = NeuroID.getInstance()?.setRegisteredUserID("Bad User REGISTERED ID")
+
+        assertEquals(false, value)
```
Comment (cometman) [inline NeuroID/src/test/java/com/neuroid/tracker/NeuroIDClassUnitTests.kt:297]: Could also add the same check as you did in the other test, just to ensure this event is not added to the queue on error `assertEquals(0, storedEvents.count())`, maybe also set the SDK to started, just to have a consistent test

**Review by cometman — CHANGES_REQUESTED** [review]

**Review by priya-xavier — APPROVED** [review]

**Review by cometman — APPROVED** [review]
lgtm!

File: NeuroID/src/test/java/com/neuroid/tracker/NeuroIDClassUnitTests.kt (hunk 6 above)
Comment (kishisaka-nid) [reply NeuroID/src/test/java/com/neuroid/tracker/NeuroIDClassUnitTests.kt:292]: fixed

**Pull request merged**